components/eamxx/src/physics/shoc/eamxx_shoc_process_interface.cpp

```diff
@@ -461,5 +465,53 @@ void SHOCMacrophysics::finalize_impl()
   // Do nothing
 }
 // =========================================================================================
+void SHOCMacrophysics::check_flux_state_consistency(const double dt)
+{
+  using PC = scream::physics::Constants<Real>;
+  const Real gravit = PC::gravit;
+  const Real qmin   = 1e-12; // minimum permitted constituent concentration (kg/kg)
+
+  const auto& pseudo_density = get_field_in ("pseudo_density").get_view<const Spack**>();
+  const auto& surf_evap      = get_field_out("surf_evap").get_view<Real*>();
+  const auto& qv             = get_field_out("qv").get_view<Spack**>();
 
+  const auto nlevs           = m_num_levs;
+  const auto nlev_packs      = ekat::npack<Spack>(nlevs);
+  const auto last_pack_idx   = (nlevs-1)/Spack::n;
+  const auto last_pack_entry = (nlevs-1)%Spack::n;
+  const auto policy          = ekat::ExeSpaceUtils<KT::ExeSpace>::get_default_team_policy(m_num_cols, nlev_packs);
+  Kokkos::parallel_for("check_flux_state_consistency",
+                       policy,
+                       KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const auto i = team.league_rank();
+
+    const auto& pseudo_density_i = ekat::subview(pseudo_density, i);
+    const auto& qv_i             = ekat::subview(qv, i);
+
+    // reciprocal of pseudo_density at the bottom layer
+    const auto rpdel = 1.0/pseudo_density_i(last_pack_idx)[last_pack_entry];
+
+    // Check if the negative surface latent heat flux can exhaust
+    // the moisture in the lowest model level. If so, apply fixer.
+    const auto condition = surf_evap(i) - (qmin - qv_i(last_pack_idx)[last_pack_entry])/(dt*gravit*rpdel);
+    if (condition < 0) {
+      const auto cc = abs(surf_evap(i)*dt*gravit);
```

Comment:
`abs` rather than `std::abs` potentially calls C's `int abs(int)` rather than, essentially, `fabs`. You might check if that's happening here. For me, this reproducer shows it's something to watch out for:
```
$ cat abs.cpp
#include <iostream>
int main () { std::cout << abs(-3.14) << "\n"; }
$ g++ -std=c++14 abs.cpp; if [ $? == 0 ]; then ./a.out; fi
3
$ cat abs.cpp
#include <iostream>
int main () { std::cout << std::abs(-3.14) << "\n"; }
$ g++ -std=c++14 abs.cpp; if [ $? == 0 ]; then ./a.out; fi
3.14
```

If I'm right about this, here are some others I'm seeing:
```
./tests/uncoupled/rrtmgp/rrtmgp_standalone_unit.cpp:209:            p_del(icol,ilay) = abs(p_lev(icol,ilay+1) - p_lev(icol,ilay));
./src/physics/nudging/tests/nudging_tests.cpp:312:          REQUIRE(abs(T_mid_v_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:313:          REQUIRE(abs(qv_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:314:          REQUIRE(abs(u_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:315:          REQUIRE(abs(v_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:328:          REQUIRE(abs(T_mid_v_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:329:          REQUIRE(abs(qv_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:330:          REQUIRE(abs(u_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:331:          REQUIRE(abs(v_h_o(icol,ilev) - val_tim_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:348:  REQUIRE(abs(T_mid_v_h_o(icol,ilev) - val_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:349:  REQUIRE(abs(qv_h_o(icol,ilev) - val_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:350:  REQUIRE(abs(u_h_o(icol,ilev) - val_avg)<0.001);
./src/physics/nudging/tests/nudging_tests.cpp:351:  REQUIRE(abs(v_h_o(icol,ilev) - val_avg)<0.001);
./src/physics/p3/impl/p3_calc_rime_density_impl.hpp:48:      Spack V_impact = abs(vtrmi1-Vt_qc);
./src/physics/shoc/tests/shoc_pdf_qw_parameters_tests.cpp:88:      REQUIRE(abs(SDS.w1_1) > abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_qw_parameters_tests.cpp:92:      REQUIRE(abs(SDS.w1_1) < abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_qw_parameters_tests.cpp:96:      REQUIRE(abs(SDS.w1_1) == abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_qw_parameters_tests.cpp:104:    Real qwgaus_diff_result1 = abs(qvconv*SDS.qw1_2 - qvconv*SDS.qw1_1);
./src/physics/shoc/tests/shoc_pdf_qw_parameters_tests.cpp:113:    Real qwgaus_diff_result2 = abs(qvconv*SDS.qw1_2 - qvconv*SDS.qw1_1);
./src/physics/shoc/tests/shoc_diag_third_tests.cpp:159:        REQUIRE(abs(SDS.w3[offset]) < 1);
./src/physics/shoc/tests/shoc_energy_integral_tests.cpp:102:          REQUIRE(abs(SDS.u_wind[offsets]) > abs(SDS.u_wind[offset]));
./src/physics/shoc/tests/shoc_energy_integral_tests.cpp:103:          REQUIRE(abs(SDS.v_wind[offsets]) > abs(SDS.v_wind[offset]));
./src/physics/shoc/tests/shoc_compute_diag_third_tests.cpp:166:        REQUIRE(abs(SDS.w3[offset]) < 10);
./src/physics/shoc/tests/shoc_clip_third_moms_tests.cpp:76:        if (abs(SDS.w3[offset]) > 1000){
./src/physics/shoc/tests/shoc_clip_third_moms_tests.cpp:94:        if (abs(w3_in[n]) > 1000){
./src/physics/shoc/tests/shoc_clip_third_moms_tests.cpp:95:          REQUIRE(abs(SDS.w3[offset]) < abs(w3_in[n]));
./src/physics/shoc/tests/shoc_pdf_thl_parameters_tests.cpp:88:      REQUIRE(abs(SDS.w1_1) > abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_thl_parameters_tests.cpp:92:      REQUIRE(abs(SDS.w1_1) < abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_thl_parameters_tests.cpp:96:      REQUIRE(abs(SDS.w1_1) == abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_thl_parameters_tests.cpp:151:    Real thlgaus_diff_result1 = abs(SDS.thl1_2 - SDS.thl1_1);
./src/physics/shoc/tests/shoc_pdf_thl_parameters_tests.cpp:160:    Real thlgaus_diff_result2 = abs(SDS.thl1_2 - SDS.thl1_1);
./src/physics/shoc/tests/shoc_pdf_computetemp_tests.cpp:61:    Int num_tests = SDS.pval/abs(presincr);
./src/physics/shoc/tests/shoc_pdf_vv_parameters_tests.cpp:95:    REQUIRE(abs(SDS.w1_1) > abs(SDS.w1_2));
./src/physics/shoc/tests/shoc_pdf_vv_parameters_tests.cpp:128:    REQUIRE(abs(SDS.w1_1) < abs(SDS.w1_2));
./src/physics/shoc/impl/shoc_assumed_pdf_impl.hpp:139:      const Smask condition =  thlsec > (thl_tol*thl_tol) && ekat::abs(w1_2 - w1_1) > w_thresh;
./src/physics/shoc/impl/shoc_assumed_pdf_impl.hpp:146:        const auto tsign = ekat::abs(tmp_val_1 - tmp_val_2);
./src/physics/shoc/impl/shoc_assumed_pdf_impl.hpp:177:      const Smask condition = qwsec > (rt_tol*rt_tol) && ekat::abs(w1_2 - w1_1) > w_thresh;
./src/physics/shoc/impl/shoc_assumed_pdf_impl.hpp:182:      const auto tsign = ekat::abs(tmp_val_1 - tmp_val_2);
./src/share/tests/common_physics_functions_tests.cpp:28:    return not ( k<NumLevels && abs(lhs-rhs)>=tol );
./src/share/tests/common_physics_functions_tests.cpp:32:    return abs(computed-expected)/abs(expected) < tol;
./src/share/tests/common_physics_functions_tests.cpp:52:    return (range_mask && abs(lhs-rhs)>=tol).none();
./src/share/tests/common_physics_functions_tests.cpp:55:    return (abs(computed-expected)/abs(expected) < tol).all();
```